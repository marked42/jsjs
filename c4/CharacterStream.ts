export class CharacterStream {
  static EOF = -1
  private index = 0
  constructor(private input: string) {}

  peek() {
    const code = this.input.charCodeAt(this.index)
    if (Number.isNaN(code)) {
      return CharacterStream.EOF
    }

    return code
  }

  next() {
    const code = this.peek()
    if (code !== CharacterStream.EOF) {
      this.index++
    }

    return code
  }

  prev() {
    this.index -= 1
  }
}
