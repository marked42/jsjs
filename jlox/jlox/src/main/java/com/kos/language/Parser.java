package com.kos.language;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.kos.language.TokenType.*;

public class Parser {
    private final List<Token> tokens;
    private int current = 0;

    Parser(List<Token> tokens) {
        this.tokens = tokens;
    }

    /**
     * expression     → assignment ;
     * assignment     → IDENTIFIER "=" assignment | logic_or ;
     * logic_or       → logic_and ( "or" logic_and )* ;
     * logic_and      → equality ( "and" equality )* ;
     * equality       → comparison ( ( "!=" | "==" ) comparison )* ;
     * comparison     → term ( ( ">" | ">=" | "<" | "<=" ) term )* ;
     * term           → factor ( ( "-" | "+" ) factor )* ;
     * factor         → unary ( ( "/" | "*" ) unary )* ;
     * unary          → ( "!" | "-" ) unary | primary ;
     * primary        → NUMBER | STRING | "true" | "false" | "nil" | "(" expression ")" ;
     */
    private Expr expression() {
        return assignment();
    }

    private Expr or() {
        Expr left = and();

        while (match(TokenType.OR)) {
            Token operator = previous();
            Expr right = and();
            left = new Expr.Logical(left, operator, right);
        }

        return left;
    }

    private Expr and() {
        Expr left = equality();

        while (match(TokenType.AND)) {
            Token operator = previous();
            Expr right = equality();
            left = new Expr.Logical(left, operator, right);
        }

        return left;
    }

    private Expr assignment() {
        Expr expr = or();

        if (match(EQUAL)) {
            Token equals = previous();
            Expr value = assignment();

            if (expr instanceof Expr.Variable) {
                Token name = ((Expr.Variable)expr).name;
                return new Expr.Assign(name, value);
            } else if (expr instanceof Expr.Get) {
                Expr.Get get = (Expr.Get)expr;
                return new Expr.Set(get.object, get.name, value);
            }

            error(equals, "Invalid assignment target.");
        }

        return expr;
    }

    List<Stmt> parse() {
        try {
            List<Stmt> statements = new ArrayList<>();
            while (!isAtEnd()) {
                statements.add(declaration());
            }
            return statements;
        } catch (ParseError error) {
            return null;
        }
    }

    /**
     * statement    → exprStmt
     *              | ifStmt
     *              | printStmt
     *              | whileStmt
     *              | returnStmt
     *              | block ;
     * whileStmt    → "while" "(" expression ")" statement ;
     */
    private Stmt statement() {
        if (match(PRINT)) return printStatement();
        if (match(LEFT_BRACE)) return new Stmt.Block(block());
        if (match(IF)) return ifStatement();
        if (match(WHILE)) return whileStatement();
        if (match(FOR)) return forStatement();
        if (match(BREAK)) return breakStatement();
        if (match(CONTINUE)) return continueStatement();
        if (match(RETURN)) return returnStatement();
        if (match(CLASS)) return classDeclaration();

        return expressionStatement();
    }

    private Stmt.Class classDeclaration() {
        Token name = consume(IDENTIFIER, "Expect class name");

        List<Stmt.Function> methods = new ArrayList<>();
        consume(LEFT_BRACE, "Expect '{' after class name");
        while (!check(TokenType.RIGHT_BRACE) && !isAtEnd()) {
            methods.add(function("method"));
        }

        consume(RIGHT_BRACE, "Expect '}' after class body");

        return new Stmt.Class(name, methods);
    }

    private Stmt.Return returnStatement() {
        Expr value = null;
        Token keyword = previous();
        if (!check(SEMICOLON)) {
            value = expression();
        }

        consume(SEMICOLON, "Expect ';' after at the end of return statement;");
        return new Stmt.Return(keyword, value);
    }

    /**
     * breakStatement通过抛出异常实现
     */
    private Stmt breakStatement() {
        consume(SEMICOLON, "Expect ';' after break.");
        return new Stmt.Break();
    }

    /**
     * continueStatement 通过抛出异常实现的话需要 forStatement 有专门的类型定义，而不是转换成while的语法糖
     */
    private Stmt continueStatement() {
        consume(SEMICOLON, "Expect ';' after continue.");
        return new Stmt.Continue();
    }

    private Stmt forStatement() {
        consume(LEFT_PAREN, "Expect '(' after for.");

        Stmt initializer;
        if (match(SEMICOLON)) {
            initializer = null;
        } else if (match(VAR)) {
            initializer = varDeclaration();
        } else {
            initializer = expressionStatement();
        }

        Expr condition;
        if (match(SEMICOLON)) {
            condition = null;
        } else {
            condition = expression();
        }
        consume(SEMICOLON, "Expect ';' after for condition.");

        Expr increment;
        if (match(RIGHT_PAREN)) {
            increment = null;
        } else {
            increment = expression();
        }
        consume(RIGHT_PAREN, "Expect ')' after for increment.");

        Stmt body = statement();

        // execute increment after for body
        body = new Stmt.Block(Arrays.asList(body, new Stmt.Expression(increment)));

        condition = condition == null ? new Expr.Literal(true) : condition;

        // transform for to while loop with proper condition
        body = new Stmt.While(condition, body);

        if (initializer != null) {
            body = new Stmt.Block(Arrays.asList( initializer, body));
        }

        return body;
    }

    private Stmt whileStatement() {
        consume(LEFT_PAREN, "Expect '(' after while.");
        Expr condition = expression();
        consume(RIGHT_PAREN, "Expect ')' after while condition");
        Stmt body = statement();

        return new Stmt.While(condition, body);
    }

    private Stmt.Condition ifStatement() {
        consume(LEFT_PAREN, "Expect '(' after if.");
        Expr condition = expression();
        consume(RIGHT_PAREN, "Expect ')' after if condition.");
        Stmt thenBranch = statement();
        Stmt elseBranch = null;
        if (match(ELSE)) {
            elseBranch = statement();
        }

        return new Stmt.Condition(condition, thenBranch, elseBranch);
    }

    private List<Stmt> block() {
        List<Stmt> statements = new ArrayList<>();

        while (!check(RIGHT_BRACE) && !isAtEnd()) {
            statements.add(declaration());
        }

        consume(RIGHT_BRACE, "Expect '}' after block");
        return statements;
    }

    /**
     *
     * declaration    → funDecl | varDecl | statement ;
     * funDecl        → "fun" function ;
     * function       → IDENTIFIER "(" parameters? ")" block ;
     * parameters     → IDENTIFIER ( "," IDENTIFIER )* ;
     */
    private Stmt declaration() {
        try {
            if (match(VAR)) return varDeclaration();
            if (match(FUN)) return function("function");

            return statement();
        } catch (ParseError error) {
            synchronize();
            return null;
        }
    }

    private Stmt.Function function(String kind) {
        Token name = consume(TokenType.IDENTIFIER, "Expect " + kind + " name");
        consume(TokenType.LEFT_PAREN, "Expect '(' after " + kind + " name");

        List<Token> params = parameters();

        consume(LEFT_BRACE, "Expect '{' before " + kind + " body.");
        List<Stmt> body = block();

        return new Stmt.Function(name, params, body);
    }

    private void checkParametersLimit(int count) {
        if (count >= 255) {
            error(peek(), "Can't have more than 255 parameters.");
        }
    }

    private List<Token> parameters() {
        List<Token> params = new ArrayList<>();

        if (!match(RIGHT_PAREN)) {
            do {
                this.checkParametersLimit(params.size());

                params.add(consume(IDENTIFIER, "Expect parameter name."));
            } while (match(COMMA));
            consume(RIGHT_PAREN, "Expect ')' after parameters.");
        }

        return params;
    }

    private Stmt varDeclaration() {
        Token name = consume(IDENTIFIER, "Expect variable name");

        Expr initializer = null;
        if (match(EQUAL)) {
            initializer = expression();
        }

        consume(SEMICOLON, "Expect ';' after variable declaration.");
        return new Stmt.Var(name, initializer);
    }


    // TODO:
    private void synchronize() {

    }

    private Stmt.Print printStatement() {
        Expr expression = expression();
        consume(SEMICOLON, "Expect ';' after value.");
        return new Stmt.Print(expression);
    }

    private Stmt.Expression expressionStatement() {
        Expr expression = expression();
        consume(SEMICOLON, "Expect ';' after expression.");
        return new Stmt.Expression(expression);
    }

    private Expr equality() {
        Expr left = comparison();

        while (match(BANG_EQUAL, EQUAL_EQUAL)) {
            Token operator = previous();
            Expr right = comparison();
            left = new Expr.Binary(left, operator, right);
        }

        return left;
    }

    private boolean match(TokenType... types) {
        for (TokenType type: types) {
            if (check(type)) {
                advance();
                return true;
            }
        }

        return false;
    }

    private boolean check(TokenType type) {
        if (isAtEnd()) return false;
        return peek().type == type;
    }

    private boolean isAtEnd() {
        return peek().type == EOF;
    }

    private Token previous() {
        return tokens.get(current - 1);
    }

    private Token advance() {
        return this.tokens.get(current++);
    }

    private Token peek() {
        return tokens.get(current);
    }

    private Expr comparison() {
        Expr left = term();

        while (match(GREATER, GREATER_EQUAL, LESS, LESS_EQUAL)) {
            Token operator = previous();
            Expr right = term();
            left = new Expr.Binary(left, operator, right);
        }

        return left;
    }

    private Expr term() {
        Expr left = factor();

        while (match(MINUS, PLUS)) {
            Token operator = previous();
            Expr right = factor();
            left = new Expr.Binary(left, operator, right);
        }

        return left;
    }

    private Expr factor() {
        Expr left = unary();

        while (match(SLASH, STAR)) {
            Token operator = previous();
            Expr right = unary();
            left = new Expr.Binary(left, operator, right);
        }
        return left;
    }

    private Expr unary() {
        if (match(BANG, MINUS)) {
            Token operator = previous();
            Expr right = unary();
            return new Expr.Unary(operator, right);
        }

        return call();
    }

    /**
     * unary          → ( "!" | "-" ) unary | call ;
     * call           → primary ( "(" arguments? ")" | "." IDENTIFIER )* ;
     */
    private Expr call() {
        Expr expr = primary();

        while (true) {
            if (match(LEFT_PAREN)) {
                expr = finishCallExpr(expr);
            } else if (match(DOT)) {
                expr = finishGetExpr(expr);
            } else {
                break;
            }
        }

        return expr;
    }

    private Expr finishCallExpr(Expr expr) {
        List<Expr> arguments = new ArrayList<>();
        if (!check(RIGHT_PAREN)) {
            do {
                this.checkParametersLimit(arguments.size());

                arguments.add(expression());
            } while (match(COMMA));
        }
        Token paren = consume(RIGHT_PAREN, "Expect ')' after arguments.");

        return new Expr.Call(expr, paren, arguments);
    }

    private Expr finishGetExpr(Expr object) {
        Token name = consume(IDENTIFIER, "Expect an identifier.");

        return new Expr.Get(object, name);
    }

    private Expr primary() {
        if (match(FALSE)) return new Expr.Literal(false);
        if (match(TRUE)) return new Expr.Literal(true);
        if (match(NIL)) return new Expr.Literal(null);
        if (match(IDENTIFIER)) return new Expr.Variable(previous());

        if (match(NUMBER, STRING)) {
            return new Expr.Literal(previous().literal);
        }

        if (match(LEFT_PAREN)) {
            Expr expr = expression();
            consume(RIGHT_PAREN, "Expect ')' after expression.");
            return new Expr.Grouping(expr);
        }

        throw new Error("expect expression");
    }

    Token consume(TokenType type, String message) {
        if (check(type)) {
            return advance();
        }

        throw error(peek(), message);
    }

    static ParseError error(Token token, String message) {
        Lox.error(token, message);
        return new ParseError();
    }

    private static class ParseError extends RuntimeException {
    }
}
