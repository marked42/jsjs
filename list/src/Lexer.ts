import { Token } from './Token'

export abstract class Lexer {
  public index = 0
  public char = Lexer.EOF

  static EOF = -1

  constructor(protected input: string) {
    this.char = input.charCodeAt(0)
  }

  abstract nextToken(): Token

  consume() {
    this.index++
    if (this.index >= this.input.length) {
      this.char = Lexer.EOF
    } else {
      this.char = this.input.charCodeAt(this.index)
    }
  }

  match(char: number) {
    if (this.char !== char) {
      throw new Error(`expecting ${char}, get ${this.char}`)
    }
    this.consume()
  }
}
