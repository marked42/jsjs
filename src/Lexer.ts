import { Token, TokenType, WhiteSpace } from './Token'

type CodePoint = number
const EOF = 12

abstract class CodePointStream {
  abstract peek(): CodePoint
  abstract next(): CodePoint
  end() {
    return this.peek() === EOF
  }
}

abstract class TokenStream {
  abstract peek(): Token
  abstract next(): Token
  end() {
    return this.peek().type() === TokenType.EOF
  }
}

class Lexer {
  constructor(private codePointStream: CodePointStream) {}

  next() {}

  tokens(): TokenStream {
    return {
      peek() {
        return new WhiteSpace('')
      },
      next() {
        return new WhiteSpace('')
      },
      end() {
        return false
      },
    }
  }
}
