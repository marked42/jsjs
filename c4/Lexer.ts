import { CharacterStream } from './CharacterStream'
import { Token, TokenType, Identifier, StringLiteral } from './Token'
import * as Errors from './Errors'

export class Lexer {
  private codes: number[] = []
  private charCode: number = -1
  constructor(private stream: CharacterStream) {}

  getChar() {
    this.charCode = this.stream.next()
    if (this.charCode !== CharacterStream.EOF) {
      this.codes.push(this.charCode)
    }
    return this.charCode
  }

  ungetChar() {
    this.codes.splice(this.codes.length - 1, 1)
    this.stream.prev()
    this.charCode = this.codes[this.codes.length - 1]
  }

  resetCharCodes() {
    this.codes = []
    this.charCode = -1
  }

  next(): Token {
    this.getChar()
    // StringLiteral
    if (charCodeIs(this.charCode, '"')) {
      const value = []

      while (true) {
        this.getChar()

        if (charCodeIs(this.charCode, '\\')) {
          this.getChar()
          if (charCodeIs(this.charCode, '"')) {
            value.push('"'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, "'")) {
            value.push("'".charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'b')) {
            value.push('\b'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'f')) {
            value.push('\f'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'n')) {
            value.push('\n'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'r')) {
            value.push('\r'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 't')) {
            value.push('\t'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'v')) {
            value.push('\v'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, '\\')) {
            value.push('\\'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, '\r')) {
            value.push('\r'.charCodeAt(0))
            this.getChar()
            if (charCodeIs(this.charCode, '\n')) {
              value.push('\n'.charCodeAt(0))
            } else {
              this.ungetChar()
            }
          } else if (charCodeIs(this.charCode, '\n')) {
            value.push('\n'.charCodeAt(0))
          } else {
            console.error(
              'invalid escape sequence \\' + String.fromCharCode(this.charCode)
            )
            value.push(this.charCode)
          }
        } else if (
          charCodeIs(this.charCode, '\r') ||
          charCodeIs(this.charCode, '\n')
        ) {
          throw Errors.StringLiteralUnescapedNewline
        } else if (charCodeIs(this.charCode, '"')) {
          const source = String.fromCharCode(...this.codes)
          this.resetCharCodes()
          return {
            type: TokenType.StringLiteral,
            source,
            value: String.fromCharCode(...value),
          } as StringLiteral
        } else if (this.charCode === CharacterStream.EOF) {
          throw Errors.StringLiteralEndEarly
        } else {
          value.push(this.charCode)
        }
      }
      // NumericLiteral
    } else if (isDigit(this.charCode)) {
      // LineComment
    } else if (charCodeIs(this.charCode, '/')) {
      this.stream.next()

      if (charCodeIs(this.stream.peek(), '/')) {
        this.stream.next()
        const codes = []
        while (true) {
          const code = this.stream.peek()
          if (!charCodeIs(code, '\r') && !charCodeIs(code, '\n')) {
            codes.push(code)
            this.stream.next()
          } else {
            return {
              type: TokenType.LineComment,
              source: String.fromCharCode(...codes),
            }
          }
        }
      } else {
        return {
          type: TokenType.Operator,
          source: '/',
        }
      }
      // Whitespace
    } else if (isWhitespace(this.charCode)) {
      const codes = [this.stream.next()]
      while (isWhitespace(this.stream.peek())) {
        codes.push(this.stream.next())
      }

      return {
        type: TokenType.Whitespace,
        source: String.fromCharCode(...codes),
      }
      // Newline
    } else if (charCodeIs(this.charCode, '\r')) {
      this.stream.next()
      if (charCodeIs(this.stream.peek(), '\n')) {
        this.stream.next()
        return {
          type: TokenType.Newline,
          source: '\r\n',
        }
      } else {
        return {
          type: TokenType.Newline,
          source: '\r',
        }
      }
    } else if (charCodeIs(this.charCode, '\n')) {
      this.stream.next()
      return {
        type: TokenType.Whitespace,
        source: '\n',
      }
    } else if (isIdentifierStart(this.charCode)) {
      const codes = [this.stream.next()]

      while (true) {
        const code = this.stream.peek()
        if (isIdentifierContinue(code)) {
          codes.push(code)
          this.stream.next()
          continue
        }

        const source = String.fromCharCode(...codes)
        return {
          type: TokenType.Identifier,
          source,
          name: source,
        } as Identifier
      }
    }

    throw new Error('Unexpected token ' + String.fromCharCode(this.charCode))
  }
}

function charCodeIs(code: number, text: string) {
  if (text.length !== 1) {
    throw new Error('text must be single character, got ' + text)
  }
  return code === text.charCodeAt(0)
}

function isDigit(code: number) {
  return code >= '0'.charCodeAt(0) && code <= '9'.charCodeAt(0)
}

function isWhitespace(code: number) {
  const ws = [' ', '\t'].map((c) => c.charCodeAt(0))

  return ws.includes(code)
}

function isLetter(code: number) {
  const isLower = code >= 'a'.charCodeAt(0) && code <= 'z'.charCodeAt(0)
  const isUpper = code >= 'A'.charCodeAt(0) && code <= 'Z'.charCodeAt(0)

  return isLower || isUpper
}

function isIdentifierStart(code: number) {
  return isLetter(code) || charCodeIs(code, '_')
}

function isIdentifierContinue(code: number) {
  return isLetter(code) || charCodeIs(code, '_') || isDigit(code)
}
