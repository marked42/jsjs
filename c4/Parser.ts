import {
  BinaryExpression,
  Expression,
  Identifier,
  MemberExpression,
  NumericLiteral,
  StringLiteral,
  UnaryExpression,
} from './AST'
import { Lexer } from './Lexer'
import { TokenType } from './Token'
import {
  getPrefixOperatorPrecedence,
  getInfixOperatorPrecedenceAssociativity,
  getPostifxOperatorPrecedenceAssociativity,
} from './Operators'

export class Parser {
  constructor(private lexer: Lexer) {
    lexer.skipComment()
    lexer.skipNewline()
    lexer.skipWhitespace()
  }

  match(type: TokenType) {
    const token = this.lexer.next()

    if (token.type !== type) {
      throw new Error(`expect token type ${type}, encountered ${token}`)
    }

    return token
  }

  // 1. 起始优先级minPrecedence默认为0
  expression(minPrecedence: number = 0): Expression {
    const token = this.lexer.next()
    let result: Expression

    // 2. 操作数生成对应的ASTNode
    if (token.type === TokenType.Identifier) {
      result = new Identifier(token.name)
    } else if (token.type === TokenType.StringLiteral) {
      result = new StringLiteral(token.value)
    } else if (token.type === TokenType.NumericLiteral) {
      result = new NumericLiteral(token.value)
    } else if (token.type === TokenType.LeftParen) {
      result = this.expression()
      result.parenthesized = true
      this.match(TokenType.RightParen)
    } else if (
      // check operators
      token.type === TokenType.Plus ||
      token.type === TokenType.Minus ||
      token.type === TokenType.Star ||
      token.type === TokenType.Div
    ) {
      // 3. 前缀操作符处理
      const precedence = getPrefixOperatorPrecedence(token.source)
      const operand = this.expression(precedence)

      result = new UnaryExpression(operand, token.source, true)
    } else {
      // 4. 其余为非法token，包括EOF
      throw new Error(`unexpected token ${token}`)
    }

    // 5. 中缀、后缀操作符的解析有三种情况
    // 5.1. 优先级低终止循环，当前调用返回
    // 5.2. 优先级高且右侧还能结合操作数，递归调用继续循环
    // 5.3. 优先级高但是右侧不能结合操作数，终止循环
    while (true) {
      const token = this.lexer.next()

      // 6. 循环终止条件一 没有更多token
      if (token.type === TokenType.EOF) {
        break
      }

      if (
        token.type === TokenType.Plus ||
        token.type === TokenType.Minus ||
        token.type === TokenType.Star ||
        token.type === TokenType.Div ||
        token.type === TokenType.StarStar ||
        token.type === TokenType.Dot
      ) {
        const {
          precedence,
          associativity,
        } = getInfixOperatorPrecedenceAssociativity(token.source)

        // 7. 二元操作符优先级低，
        if (precedence < minPrecedence) {
          this.lexer.prev()
          break
        }

        const newLp = associativity === 'left' ? precedence + 1 : precedence
        const right = this.expression(newLp)
        // TODO: binary operator type warning
        if (token.type === TokenType.Dot) {
          result = new MemberExpression(result, right)
        } else {
          result = new BinaryExpression(result, right, token.source as any)
        }
        continue
      }

      if (
        token.type === TokenType.Increment ||
        token.type === TokenType.Decrement
      ) {
        const precedence = getPostifxOperatorPrecedenceAssociativity(
          token.source
        )
        if (precedence > minPrecedence) {
          result = new UnaryExpression(result, token.source, false)
          // TODO: 也可以使用peek操作， peek和prev是实现同样功能的两种方式，都需要token缓冲区支持
        } else {
          this.lexer.prev()
        }
        break
      }

      if (token.type === TokenType.RightParen) {
        this.lexer.prev()
        break
      }

      if (token.type === TokenType.LeftSquare) {
        const {
          precedence,
          associativity,
        } = getInfixOperatorPrecedenceAssociativity(token.source)

        // 7. 二元操作符优先级低，
        if (precedence < minPrecedence) {
          this.lexer.prev()
          break
        }

        const member = this.expression()
        this.match(TokenType.RightSquare)
        result = new MemberExpression(result, member, true)
        continue
      }

      if (token.type === TokenType.RightSquare) {
        this.lexer.prev()
        break
      }

      throw new Error('unexpected token ' + JSON.stringify(token, null, 2))
    }

    return result
  }

  program() {}
}
