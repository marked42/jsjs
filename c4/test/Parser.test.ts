import { Parser } from '../Parser'
import { Lexer } from '../Lexer'
import { CharacterStream } from '../CharacterStream'

function parserForInput(input: string) {
  const stream = new CharacterStream(input)
  const lexer = new Lexer(stream)
  const parser = new Parser(lexer)

  return parser
}

describe('expression', () => {
  it('should parse binary operators with correct precedence', () => {
    const input = 'a + 1 * 2'
    const parser = parserForInput(input)

    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse left associative binary operators correctly', () => {
    const input = 'a + 1 + 2'
    const parser = parserForInput(input)
    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse right associative binary operators correctly', () => {
    const input = 'a ** 1 ** 2'
    const parser = parserForInput(input)
    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse prefix operators correctly', () => {
    const input = '-a + 1 * 2'
    const parser = parserForInput(input)
    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse prefix operators correctly at right', () => {
    const input = '1 + -a'
    const parser = parserForInput(input)
    expect(parser.expression()).toMatchSnapshot()
  })

  it('should parse postfix operators correctly', () => {
    const input = '-a++ + 1 * 2'
    const parser = parserForInput(input)
    expect(parser.expression()).toMatchSnapshot()
  })
})
