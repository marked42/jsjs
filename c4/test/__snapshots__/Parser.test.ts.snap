// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expression should parse binary operators with correct precedence 1`] = `
BinaryExpression {
  "left": Identifier {
    "name": "a",
    "parenthesized": false,
    "type": "Identifier",
  },
  "op": "+",
  "parenthesized": false,
  "right": BinaryExpression {
    "left": NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 1,
    },
    "op": "*",
    "parenthesized": false,
    "right": NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 2,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`expression should parse left associative binary operators correctly 1`] = `
BinaryExpression {
  "left": BinaryExpression {
    "left": Identifier {
      "name": "a",
      "parenthesized": false,
      "type": "Identifier",
    },
    "op": "+",
    "parenthesized": false,
    "right": NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 1,
    },
    "type": "BinaryExpression",
  },
  "op": "+",
  "parenthesized": false,
  "right": NumericLiteral {
    "parenthesized": false,
    "type": "NumericLiteral",
    "value": 2,
  },
  "type": "BinaryExpression",
}
`;

exports[`expression should parse parenthesis correctly 1`] = `
BinaryExpression {
  "left": NumericLiteral {
    "parenthesized": false,
    "type": "NumericLiteral",
    "value": 1,
  },
  "op": "+",
  "parenthesized": false,
  "right": BinaryExpression {
    "left": BinaryExpression {
      "left": NumericLiteral {
        "parenthesized": false,
        "type": "NumericLiteral",
        "value": 2,
      },
      "op": "+",
      "parenthesized": true,
      "right": NumericLiteral {
        "parenthesized": false,
        "type": "NumericLiteral",
        "value": 3,
      },
      "type": "BinaryExpression",
    },
    "op": "*",
    "parenthesized": false,
    "right": NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 4,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`expression should parse postfix operators correctly 1`] = `
BinaryExpression {
  "left": UnaryExpression {
    "op": "-",
    "operand": UnaryExpression {
      "op": "++",
      "operand": Identifier {
        "name": "a",
        "parenthesized": false,
        "type": "Identifier",
      },
      "parenthesized": false,
      "prefix": false,
      "type": "UnaryExpression",
    },
    "parenthesized": false,
    "prefix": true,
    "type": "UnaryExpression",
  },
  "op": "+",
  "parenthesized": false,
  "right": BinaryExpression {
    "left": NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 1,
    },
    "op": "*",
    "parenthesized": false,
    "right": NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 2,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`expression should parse prefix operators correctly 1`] = `
BinaryExpression {
  "left": UnaryExpression {
    "op": "-",
    "operand": Identifier {
      "name": "a",
      "parenthesized": false,
      "type": "Identifier",
    },
    "parenthesized": false,
    "prefix": true,
    "type": "UnaryExpression",
  },
  "op": "+",
  "parenthesized": false,
  "right": BinaryExpression {
    "left": NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 1,
    },
    "op": "*",
    "parenthesized": false,
    "right": NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 2,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`expression should parse prefix operators correctly at right 1`] = `
BinaryExpression {
  "left": NumericLiteral {
    "parenthesized": false,
    "type": "NumericLiteral",
    "value": 1,
  },
  "op": "+",
  "parenthesized": false,
  "right": UnaryExpression {
    "op": "-",
    "operand": Identifier {
      "name": "a",
      "parenthesized": false,
      "type": "Identifier",
    },
    "parenthesized": false,
    "prefix": true,
    "type": "UnaryExpression",
  },
  "type": "BinaryExpression",
}
`;

exports[`expression should parse right associative binary operators correctly 1`] = `
BinaryExpression {
  "left": Identifier {
    "name": "a",
    "parenthesized": false,
    "type": "Identifier",
  },
  "op": "**",
  "parenthesized": false,
  "right": BinaryExpression {
    "left": NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 1,
    },
    "op": "**",
    "parenthesized": false,
    "right": NumericLiteral {
      "parenthesized": false,
      "type": "NumericLiteral",
      "value": 2,
    },
    "type": "BinaryExpression",
  },
  "type": "BinaryExpression",
}
`;
